Serialization/Deserialization issue with JSON

I am trying to switch to version 1.0.2 and just noticed an issue with Jackson serialization/deserialization. 
The problem comes from the usage of wildcard recently introduce in collections.

For example in `Ehr` class:
```java
 private List<ObjectRef<?>> contributions = new ArrayList<>();
```

Here is the test case to reproduce the issue:
```
public class EhrTest {

    @Test
    public void testJsonSerialization() throws IOException {
        Ehr expected = new Ehr();

        List<ObjectRef<?>> contributions = new ArrayList<>();
        contributions.add(new ObjectRef<>(new HierObjectId("value"), "namespace", "type"));
        expected.setContributions(contributions);

        StringWriter sw = new StringWriter();
        ObjectMapper objectMapper = JacksonUtil.getObjectMapper(RMJacksonConfiguration.createStandardsCompliant());
        objectMapper.writeValue(sw, expected);

        Ehr actual = objectMapper.readValue(sw.toString(), Ehr.class);
        assertEquals(expected, actual);
    }
}
```

Output:
```
Problem deserializing property 'id' (expected type: [simple type, class java.lang.Object]; actual type: `java.util.LinkedHashMap`), problem: argument type mismatch
 at [Source: (String)"{
  "contributions" : [ {
    "_type" : "OBJECT_REF",
    "namespace" : "namespace",
    "type" : "type",
    "id" : {
      "value" : "value"
    }
  } ],
  "compositions" : [ ],
  "folders" : [ ]
}"; line: 8, column: 5] (through reference chain: com.nedap.archie.rm.ehr.Ehr["contributions"]->java.util.ArrayList[0]->com.nedap.archie.rm.support.identification.ObjectRef["id"])
com.fasterxml.jackson.databind.JsonMappingException: Problem deserializing property 'id' (expected type: [simple type, class java.lang.Object]; actual type: `java.util.LinkedHashMap`), problem: argument type mismatch
 at [Source: (String)"{
  "contributions" : [ {
    "_type" : "OBJECT_REF",
    "namespace" : "namespace",
    "type" : "type",
    "id" : {
      "value" : "value"
    }
  } ],
  "compositions" : [ ],
  "folders" : [ ]
}"; line: 8, column: 5] (through reference chain: com.nedap.archie.rm.ehr.Ehr["contributions"]->java.util.ArrayList[0]->com.nedap.archie.rm.support.identification.ObjectRef["id"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:274)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:609)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:143)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:225)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:197)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:139)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:107)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeWithType(BeanDeserializerBase.java:1292)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:357)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4593)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3548)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3516)
	at com.nedap.archie.rm.ehr.EhrTest.testJsonSerialization(EhrTest.java:31)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy5.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: argument type mismatch
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:141)
	... 67 more
```

I think that archie should be able to deserialize a JSON generated by itself.

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/Contribution.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/Contribution.java
@@ -3,6 +3,7 @@ package com.nedap.archie.rm.changecontrol;
 import com.nedap.archie.rm.RMObject;
 import com.nedap.archie.rm.generic.AuditDetails;
 import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 
 import javax.xml.bind.annotation.XmlAccessType;
@@ -24,13 +25,13 @@ import java.util.Objects;
 public class Contribution extends RMObject {
 
     private HierObjectId uid;
-    private List<ObjectRef<?>> versions = new ArrayList<>();
+    private List<ObjectRef<? extends ObjectId>> versions = new ArrayList<>();
     private AuditDetails audit;
 
     public Contribution() {
     }
 
-    public Contribution(HierObjectId uid, List<ObjectRef<?>> versions, AuditDetails audit) {
+    public Contribution(HierObjectId uid, List<ObjectRef<? extends ObjectId>> versions, AuditDetails audit) {
         this.uid = uid;
         this.versions = versions;
         this.audit = audit;
@@ -44,11 +45,11 @@ public class Contribution extends RMObject {
         this.uid = uid;
     }
 
-    public List<ObjectRef<?>> getVersions() {
+    public List<ObjectRef<? extends ObjectId>> getVersions() {
         return versions;
     }
 
-    public void setVersions(List<ObjectRef<?>> versions) {
+    public void setVersions(List<ObjectRef<? extends ObjectId>> versions) {
         this.versions = versions;
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/ImportedVersion.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/ImportedVersion.java
@@ -2,6 +2,7 @@ package com.nedap.archie.rm.changecontrol;
 
 import com.nedap.archie.rm.datavalues.DvCodedText;
 import com.nedap.archie.rm.generic.AuditDetails;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.ObjectVersionId;
 
@@ -20,7 +21,7 @@ public class ImportedVersion<T> extends Version<T> {
     public ImportedVersion() {
     }
 
-    public ImportedVersion(AuditDetails commitAudit, ObjectRef<?> contribution, @Nullable String signature, OriginalVersion<T> item) {
+    public ImportedVersion(AuditDetails commitAudit, ObjectRef<? extends ObjectId> contribution, @Nullable String signature, OriginalVersion<T> item) {
         super(commitAudit, contribution, signature);
         this.item = item;
     }

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/OriginalVersion.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/OriginalVersion.java
@@ -5,6 +5,7 @@ import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.nedap.archie.rm.datavalues.DvCodedText;
 import com.nedap.archie.rm.generic.Attestation;
 import com.nedap.archie.rm.generic.AuditDetails;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.ObjectVersionId;
 import com.nedap.archie.rminfo.Invariant;
@@ -53,7 +54,7 @@ public class    OriginalVersion<Type> extends Version<Type> {
     public OriginalVersion() {
     }
 
-    public OriginalVersion(ObjectVersionId uid, @Nullable ObjectVersionId precedingVersionUid, @Nullable Type data, DvCodedText lifecycleState, AuditDetails commitAudit, ObjectRef<?> contribution, @Nullable String signature, @Nullable List<ObjectVersionId> otherInputVersionUids, @Nullable List<Attestation> attestations) {
+    public OriginalVersion(ObjectVersionId uid, @Nullable ObjectVersionId precedingVersionUid, @Nullable Type data, DvCodedText lifecycleState, AuditDetails commitAudit, ObjectRef<? extends ObjectId> contribution, @Nullable String signature, @Nullable List<ObjectVersionId> otherInputVersionUids, @Nullable List<Attestation> attestations) {
         super(commitAudit, contribution, signature);
         this.uid = uid;
         this.precedingVersionUid = precedingVersionUid;

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/Version.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/Version.java
@@ -3,6 +3,7 @@ package com.nedap.archie.rm.changecontrol;
 import com.nedap.archie.rm.RMObject;
 import com.nedap.archie.rm.datavalues.DvCodedText;
 import com.nedap.archie.rm.generic.AuditDetails;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.ObjectVersionId;
 import com.nedap.archie.rminfo.Invariant;
@@ -29,7 +30,7 @@ import java.util.Objects;
 
 })
 public abstract class Version<Type> extends RMObject {
-    private ObjectRef<?> contribution;
+    private ObjectRef<? extends ObjectId> contribution;
     @Nullable
 
     private String signature;
@@ -39,17 +40,17 @@ public abstract class Version<Type> extends RMObject {
     public Version() {
     }
 
-    public Version(AuditDetails commitAudit, ObjectRef<?> contribution, @Nullable String signature) {
+    public Version(AuditDetails commitAudit, ObjectRef<? extends ObjectId> contribution, @Nullable String signature) {
         this.contribution = contribution;
         this.signature = signature;
         this.commitAudit = commitAudit;
     }
 
-    public ObjectRef<?> getContribution() {
+    public ObjectRef<? extends ObjectId> getContribution() {
         return contribution;
     }
 
-    public void setContribution(ObjectRef<?> contribution) {
+    public void setContribution(ObjectRef<? extends ObjectId> contribution) {
         this.contribution = contribution;
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/VersionedObject.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/changecontrol/VersionedObject.java
@@ -4,6 +4,7 @@ import com.fasterxml.jackson.annotation.JsonAlias;
 import com.nedap.archie.rm.RMObject;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 
 import javax.xml.bind.annotation.XmlAccessType;
@@ -23,7 +24,7 @@ import java.util.Objects;
 public class VersionedObject<Type> extends RMObject {
     private HierObjectId uid;
     @XmlElement(name = "owner_id")
-    private ObjectRef<?> ownerId;
+    private ObjectRef<? extends ObjectId> ownerId;
 
     @XmlElement(name = "time_created")
     private DvDateTime timeCreated;
@@ -31,7 +32,7 @@ public class VersionedObject<Type> extends RMObject {
     public VersionedObject() {
     }
 
-    public VersionedObject(HierObjectId uid, ObjectRef<?> ownerId, DvDateTime timeCreated) {
+    public VersionedObject(HierObjectId uid, ObjectRef<? extends ObjectId> ownerId, DvDateTime timeCreated) {
         this.uid = uid;
         this.ownerId = ownerId;
         this.timeCreated = timeCreated;
@@ -45,11 +46,11 @@ public class VersionedObject<Type> extends RMObject {
         this.uid = uid;
     }
 
-    public ObjectRef<?> getOwnerId() {
+    public ObjectRef<? extends ObjectId> getOwnerId() {
         return ownerId;
     }
 
-    public void setOwnerId(ObjectRef<?> ownerId) {
+    public void setOwnerId(ObjectRef<? extends ObjectId> ownerId) {
         this.ownerId = ownerId;
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Action.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Action.java
@@ -1,7 +1,6 @@
 package com.nedap.archie.rm.composition;
 
 
-import com.nedap.archie.rm.composition.IsmTransition;
 import com.nedap.archie.rm.archetyped.Archetyped;
 import com.nedap.archie.rm.archetyped.FeederAudit;
 import com.nedap.archie.rm.archetyped.Link;
@@ -13,6 +12,7 @@ import com.nedap.archie.rm.datavalues.DvText;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.generic.Participation;
 import com.nedap.archie.rm.generic.PartyProxy;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 
@@ -45,7 +45,7 @@ public class Action extends CareEntry {
     public Action() {
     }
 
-    public Action(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<?> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<?> guidelineId, DvDateTime time, ItemStructure description, IsmTransition ismTransition, @Nullable InstructionDetails instructionDetails) {
+    public Action(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<? extends ObjectId> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<? extends ObjectId> guidelineId, DvDateTime time, ItemStructure description, IsmTransition ismTransition, @Nullable InstructionDetails instructionDetails) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName, language, encoding, subject, provider, workflowId, otherParticipations, protocol, guidelineId);
         this.time = time;
         this.description = description;

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/composition/AdminEntry.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/composition/AdminEntry.java
@@ -9,6 +9,7 @@ import com.nedap.archie.rm.datatypes.CodePhrase;
 import com.nedap.archie.rm.datavalues.DvText;
 import com.nedap.archie.rm.generic.Participation;
 import com.nedap.archie.rm.generic.PartyProxy;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 
@@ -32,7 +33,7 @@ public class AdminEntry extends Entry {
     public AdminEntry() {
     }
 
-    public AdminEntry(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<?> workflowId, @Nullable List<Participation> otherParticipations, ItemStructure data) {
+    public AdminEntry(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<? extends ObjectId> workflowId, @Nullable List<Participation> otherParticipations, ItemStructure data) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName, language, encoding, subject, provider, workflowId, otherParticipations);
         this.data = data;
     }

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/composition/CareEntry.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/composition/CareEntry.java
@@ -9,6 +9,7 @@ import com.nedap.archie.rm.datatypes.CodePhrase;
 import com.nedap.archie.rm.datavalues.DvText;
 import com.nedap.archie.rm.generic.Participation;
 import com.nedap.archie.rm.generic.PartyProxy;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 
@@ -34,12 +35,12 @@ public abstract class CareEntry extends Entry {
     private ItemStructure protocol;
     @Nullable
     @XmlElement(name = "guideline_id")
-    private ObjectRef<?> guidelineId;
+    private ObjectRef<? extends ObjectId> guidelineId;
 
     public CareEntry() {
     }
 
-    public CareEntry(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<?> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<?> guidelineId) {
+    public CareEntry(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<? extends ObjectId> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<? extends ObjectId> guidelineId) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName, language, encoding, subject, provider, workflowId, otherParticipations);
         this.protocol = protocol;
         this.guidelineId = guidelineId;
@@ -56,11 +57,11 @@ public abstract class CareEntry extends Entry {
     }
 
     @Nullable
-    public ObjectRef<?> getGuidelineId() {
+    public ObjectRef<? extends ObjectId> getGuidelineId() {
         return guidelineId;
     }
 
-    public void setGuidelineId(@Nullable ObjectRef<?> guidelineId) {
+    public void setGuidelineId(@Nullable ObjectRef<? extends ObjectId> guidelineId) {
         this.guidelineId = guidelineId;
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Entry.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Entry.java
@@ -10,6 +10,7 @@ import com.nedap.archie.rm.datavalues.DvText;
 import com.nedap.archie.rm.generic.Participation;
 import com.nedap.archie.rm.generic.PartyProxy;
 import com.nedap.archie.rm.generic.PartySelf;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 import com.nedap.archie.rminfo.Invariant;
@@ -43,7 +44,7 @@ public abstract class Entry extends ContentItem {
     @Nullable
 
     @XmlElement(name = "workflow_id")
-    private ObjectRef<?> workflowId;
+    private ObjectRef<? extends ObjectId> workflowId;
     private PartyProxy subject;
     @Nullable
     private PartyProxy provider;
@@ -55,7 +56,7 @@ public abstract class Entry extends ContentItem {
     public Entry() {
     }
 
-    public Entry(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<?> workflowId, @Nullable List<Participation> otherParticipations) {
+    public Entry(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<? extends ObjectId> workflowId, @Nullable List<Participation> otherParticipations) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName);
         this.language = language;
         this.encoding = encoding;
@@ -111,12 +112,12 @@ public abstract class Entry extends ContentItem {
     }
 
     @Nullable
-    public ObjectRef<?> getWorkflowId() {
+    public ObjectRef<? extends ObjectId> getWorkflowId() {
         return workflowId;
     }
 
     @JsonAlias({"work_flow_id"})
-    public void setWorkflowId(@Nullable ObjectRef<?> workflowId) {
+    public void setWorkflowId(@Nullable ObjectRef<? extends ObjectId> workflowId) {
         this.workflowId = workflowId;
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Evaluation.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Evaluation.java
@@ -9,6 +9,7 @@ import com.nedap.archie.rm.datatypes.CodePhrase;
 import com.nedap.archie.rm.datavalues.DvText;
 import com.nedap.archie.rm.generic.Participation;
 import com.nedap.archie.rm.generic.PartyProxy;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 
@@ -34,7 +35,7 @@ public class Evaluation extends CareEntry {
     public Evaluation() {
     }
 
-    public Evaluation(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<?> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<?> guidelineId, ItemStructure data) {
+    public Evaluation(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<? extends ObjectId> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<? extends ObjectId> guidelineId, ItemStructure data) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName, language, encoding, subject, provider, workflowId, otherParticipations, protocol, guidelineId);
         this.data = data;
     }

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Instruction.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Instruction.java
@@ -12,6 +12,7 @@ import com.nedap.archie.rm.datavalues.encapsulated.DvParsable;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.generic.Participation;
 import com.nedap.archie.rm.generic.PartyProxy;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 import com.nedap.archie.rminfo.Invariant;
@@ -52,7 +53,7 @@ public class Instruction extends CareEntry {
     public Instruction() {
     }
 
-    public Instruction(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<?> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<?> guidelineId, DvText narrative, @Nullable List<Activity> activities, @Nullable DvDateTime expiryTime, @Nullable DvParsable wfDefinition) {
+    public Instruction(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<? extends ObjectId> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<? extends ObjectId> guidelineId, DvText narrative, @Nullable List<Activity> activities, @Nullable DvDateTime expiryTime, @Nullable DvParsable wfDefinition) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName, language, encoding, subject, provider, workflowId, otherParticipations, protocol, guidelineId);
         this.narrative = narrative;
         this.expiryTime = expiryTime;

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Observation.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/composition/Observation.java
@@ -10,6 +10,7 @@ import com.nedap.archie.rm.datatypes.CodePhrase;
 import com.nedap.archie.rm.datavalues.DvText;
 import com.nedap.archie.rm.generic.Participation;
 import com.nedap.archie.rm.generic.PartyProxy;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 
@@ -40,7 +41,7 @@ public class Observation extends CareEntry {
     public Observation() {
     }
 
-    public Observation(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<?> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<?> guidelineId, History<ItemStructure> data, @Nullable History<ItemStructure> state) {
+    public Observation(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, CodePhrase language, CodePhrase encoding, PartyProxy subject, @Nullable PartyProxy provider, @Nullable ObjectRef<? extends ObjectId> workflowId, @Nullable List<Participation> otherParticipations, @Nullable ItemStructure protocol, @Nullable ObjectRef<? extends ObjectId> guidelineId, History<ItemStructure> data, @Nullable History<ItemStructure> state) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName, language, encoding, subject, provider, workflowId, otherParticipations, protocol, guidelineId);
         this.state = state;
         this.data = data;

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/directory/Folder.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/directory/Folder.java
@@ -1,10 +1,6 @@
 package com.nedap.archie.rm.directory;
 
-import com.nedap.archie.rm.archetyped.Archetyped;
-import com.nedap.archie.rm.archetyped.FeederAudit;
-import com.nedap.archie.rm.archetyped.Link;
-import com.nedap.archie.rm.archetyped.Locatable;
-import com.nedap.archie.rm.archetyped.Pathable;
+import com.nedap.archie.rm.archetyped.*;
 import com.nedap.archie.rm.datastructures.ItemStructure;
 import com.nedap.archie.rm.datavalues.DvText;
 import com.nedap.archie.rm.support.identification.ObjectId;
@@ -38,7 +34,7 @@ public class Folder extends Locatable {
     private List<Folder> folders = new ArrayList<>();
 
     @Nullable
-    @XmlElement(name = "details")
+    @XmlElement(name="details")
     private ItemStructure details;
 
     public Folder() {
@@ -112,7 +108,7 @@ public class Folder extends Locatable {
         return Objects.hash(super.hashCode(), items, folders, details);
     }
 
-    @Invariant(value = "Folders_valid", ignored = true)
+    @Invariant(value="Folders_valid", ignored = true)
     public boolean foldersValid() {
         return InvariantUtil.nullOrNotEmpty(folders);
     }

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/directory/Folder.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/directory/Folder.java
@@ -1,8 +1,13 @@
 package com.nedap.archie.rm.directory;
 
-import com.nedap.archie.rm.archetyped.*;
+import com.nedap.archie.rm.archetyped.Archetyped;
+import com.nedap.archie.rm.archetyped.FeederAudit;
+import com.nedap.archie.rm.archetyped.Link;
+import com.nedap.archie.rm.archetyped.Locatable;
+import com.nedap.archie.rm.archetyped.Pathable;
 import com.nedap.archie.rm.datastructures.ItemStructure;
 import com.nedap.archie.rm.datavalues.DvText;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rm.support.identification.UIDBasedId;
 import com.nedap.archie.rminfo.Invariant;
@@ -28,25 +33,25 @@ import java.util.Objects;
 })
 public class Folder extends Locatable {
     @Nullable
-    private List<ObjectRef<?>> items = new ArrayList<>();
+    private List<ObjectRef<? extends ObjectId>> items = new ArrayList<>();
     @Nullable
     private List<Folder> folders = new ArrayList<>();
 
     @Nullable
-    @XmlElement(name="details")
+    @XmlElement(name = "details")
     private ItemStructure details;
 
     public Folder() {
     }
 
-    public Folder(String archetypeNodeId, DvText name, ItemStructure details, @Nullable List<ObjectRef<?>> items, @Nullable List<Folder> folders) {
+    public Folder(String archetypeNodeId, DvText name, ItemStructure details, @Nullable List<ObjectRef<? extends ObjectId>> items, @Nullable List<Folder> folders) {
         super(archetypeNodeId, name);
         this.items = items;
         this.folders = folders;
         this.details = details;
     }
 
-    public Folder(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, ItemStructure details, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, @Nullable List<ObjectRef<?>> items, @Nullable List<Folder> folders) {
+    public Folder(@Nullable UIDBasedId uid, String archetypeNodeId, DvText name, ItemStructure details, @Nullable Archetyped archetypeDetails, @Nullable FeederAudit feederAudit, @Nullable List<Link> links, @Nullable Pathable parent, @Nullable String parentAttributeName, @Nullable List<ObjectRef<? extends ObjectId>> items, @Nullable List<Folder> folders) {
         super(uid, archetypeNodeId, name, archetypeDetails, feederAudit, links, parent, parentAttributeName);
         this.items = items;
         this.folders = folders;
@@ -54,15 +59,15 @@ public class Folder extends Locatable {
     }
 
     @Nullable
-    public List<ObjectRef<?>> getItems() {
+    public List<ObjectRef<? extends ObjectId>> getItems() {
         return items;
     }
 
-    public void setItems(@Nullable List<ObjectRef<?>> items) {
+    public void setItems(@Nullable List<ObjectRef<? extends ObjectId>> items) {
         this.items = items;
     }
 
-    public void addItem(ObjectRef<?> item) {
+    public void addItem(ObjectRef<? extends ObjectId> item) {
         this.items.add(item);
     }
 
@@ -107,7 +112,7 @@ public class Folder extends Locatable {
         return Objects.hash(super.hashCode(), items, folders, details);
     }
 
-    @Invariant(value="Folders_valid", ignored = true)
+    @Invariant(value = "Folders_valid", ignored = true)
     public boolean foldersValid() {
         return InvariantUtil.nullOrNotEmpty(folders);
     }

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/directory/VersionedFolder.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/directory/VersionedFolder.java
@@ -3,6 +3,7 @@ package com.nedap.archie.rm.directory;
 import com.nedap.archie.rm.changecontrol.VersionedObject;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 
 import javax.xml.bind.annotation.XmlType;
@@ -16,7 +17,7 @@ public class VersionedFolder extends VersionedObject<Folder> {
     public VersionedFolder() {
     }
 
-    public VersionedFolder(HierObjectId uid, ObjectRef<?> ownerId, DvDateTime timeCreated) {
+    public VersionedFolder(HierObjectId uid, ObjectRef<? extends ObjectId> ownerId, DvDateTime timeCreated) {
         super(uid, ownerId, timeCreated);
     }
 }

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/Ehr.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/Ehr.java
@@ -41,14 +41,14 @@ public class Ehr extends RMObject {
 
     private List<ObjectRef<? extends ObjectId>> contributions = new ArrayList<>();
     @XmlElement(name="ehr_status")
-    private ObjectRef<?> ehrStatus;
+    private ObjectRef<? extends ObjectId> ehrStatus;
     @XmlElement(name="ehr_access")
-    private ObjectRef<?> ehrAccess;
+    private ObjectRef<? extends ObjectId> ehrAccess;
     @Nullable
     private List<ObjectRef<? extends ObjectId>> compositions = new ArrayList<>();
 
     @Nullable
-    private ObjectRef<?> directory;
+    private ObjectRef<? extends ObjectId> directory;
 
     @Nullable
     private List<ObjectRef<? extends ObjectId>> folders = new ArrayList<>();
@@ -59,7 +59,7 @@ public class Ehr extends RMObject {
     public Ehr() {
     }
 
-    public Ehr(HierObjectId systemId, HierObjectId ehrId, DvDateTime timeCreated, List<ObjectRef<? extends ObjectId>> contributions, ObjectRef<?> ehrStatus, ObjectRef<?> ehrAccess, @Nullable ObjectRef<?> directory, @Nullable List<ObjectRef<?>> compositions) {
+    public Ehr(HierObjectId systemId, HierObjectId ehrId, DvDateTime timeCreated, List<ObjectRef<? extends ObjectId>> contributions, ObjectRef<? extends ObjectId> ehrStatus, ObjectRef<? extends ObjectId> ehrAccess, @Nullable ObjectRef<? extends ObjectId> directory, @Nullable List<ObjectRef<? extends ObjectId>> compositions) {
         this.systemId = systemId;
         this.ehrId = ehrId;
         this.contributions = contributions;
@@ -99,19 +99,19 @@ public class Ehr extends RMObject {
         this.contributions.add(contribution);
     }
 
-    public ObjectRef<?> getEhrStatus() {
+    public ObjectRef<? extends ObjectId> getEhrStatus() {
         return ehrStatus;
     }
 
-    public void setEhrStatus(ObjectRef<?> ehrStatus) {
+    public void setEhrStatus(ObjectRef<? extends ObjectId> ehrStatus) {
         this.ehrStatus = ehrStatus;
     }
 
-    public ObjectRef<?> getEhrAccess() {
+    public ObjectRef<? extends ObjectId> getEhrAccess() {
         return ehrAccess;
     }
 
-    public void setEhrAccess(ObjectRef<?> ehrAccess) {
+    public void setEhrAccess(ObjectRef<? extends ObjectId> ehrAccess) {
         this.ehrAccess = ehrAccess;
     }
 
@@ -129,11 +129,11 @@ public class Ehr extends RMObject {
     }
 
     @Nullable
-    public ObjectRef<?> getDirectory() {
+    public ObjectRef<? extends ObjectId> getDirectory() {
         return directory;
     }
 
-    public void setDirectory(@Nullable ObjectRef<?> directory) {
+    public void setDirectory(@Nullable ObjectRef<? extends ObjectId> directory) {
         this.directory = directory;
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/Ehr.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/Ehr.java
@@ -3,12 +3,17 @@ package com.nedap.archie.rm.ehr;
 import com.nedap.archie.rm.RMObject;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.rminfo.Invariant;
 import com.nedap.archie.rmutil.InvariantUtil;
 
 import javax.annotation.Nullable;
-import javax.xml.bind.annotation.*;
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlType;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Objects;
@@ -18,7 +23,7 @@ import java.util.Objects;
  * in some kind of persistence framework. But it can be useful for transferring information
  * Created by pieter.bos on 08/07/16.
  */
-@XmlRootElement(name="EHR")
+@XmlRootElement(name = "EHR")
 @XmlType(name = "EHR", propOrder = {
         "systemId",
         "ehrId",
@@ -33,32 +38,32 @@ import java.util.Objects;
 @XmlAccessorType(XmlAccessType.FIELD)
 public class Ehr extends RMObject {
 
-    @XmlElement(name="system_id")
+    @XmlElement(name = "system_id")
     private HierObjectId systemId;
-    @XmlElement(name="ehr_id")
+    @XmlElement(name = "ehr_id")
     private HierObjectId ehrId;
 
-    private List<ObjectRef<?>> contributions = new ArrayList<>();
-    @XmlElement(name="ehr_status")
+    private List<ObjectRef<? extends ObjectId>> contributions = new ArrayList<>();
+    @XmlElement(name = "ehr_status")
     private ObjectRef<?> ehrStatus;
-    @XmlElement(name="ehr_access")
+    @XmlElement(name = "ehr_access")
     private ObjectRef<?> ehrAccess;
     @Nullable
-    private List<ObjectRef<?>> compositions = new ArrayList<>();
+    private List<ObjectRef<? extends ObjectId>> compositions = new ArrayList<>();
 
     @Nullable
     private ObjectRef<?> directory;
 
     @Nullable
-    private List<ObjectRef<?>> folders = new ArrayList<>();
+    private List<ObjectRef<? extends ObjectId>> folders = new ArrayList<>();
 
-    @XmlElement(name="time_created")
+    @XmlElement(name = "time_created")
     private DvDateTime timeCreated;
 
     public Ehr() {
     }
 
-    public Ehr(HierObjectId systemId, HierObjectId ehrId, DvDateTime timeCreated, List<ObjectRef<?>> contributions, ObjectRef<?> ehrStatus, ObjectRef<?> ehrAccess, @Nullable ObjectRef<?> directory, @Nullable List<ObjectRef<?>> compositions) {
+    public Ehr(HierObjectId systemId, HierObjectId ehrId, DvDateTime timeCreated, List<ObjectRef<? extends ObjectId>> contributions, ObjectRef<?> ehrStatus, ObjectRef<?> ehrAccess, @Nullable ObjectRef<?> directory, @Nullable List<ObjectRef<?>> compositions) {
         this.systemId = systemId;
         this.ehrId = ehrId;
         this.contributions = contributions;
@@ -86,15 +91,15 @@ public class Ehr extends RMObject {
     }
 
     @Nullable
-    public List<ObjectRef<?>> getContributions() {
+    public List<ObjectRef<? extends ObjectId>> getContributions() {
         return contributions;
     }
 
-    public void setContributions(@Nullable List<ObjectRef<?>> contributions) {
+    public void setContributions(@Nullable List<ObjectRef<? extends ObjectId>> contributions) {
         this.contributions = contributions;
     }
 
-    public void addContribution(ObjectRef<?> contribution) {
+    public void addContribution(ObjectRef<? extends ObjectId> contribution) {
         this.contributions.add(contribution);
     }
 
@@ -115,15 +120,15 @@ public class Ehr extends RMObject {
     }
 
     @Nullable
-    public List<ObjectRef<?>> getCompositions() {
+    public List<ObjectRef<? extends ObjectId>> getCompositions() {
         return compositions;
     }
 
-    public void setCompositions(@Nullable List<ObjectRef<?>> compositions) {
+    public void setCompositions(@Nullable List<ObjectRef<? extends ObjectId>> compositions) {
         this.compositions = compositions;
     }
 
-    public void addComposition(ObjectRef<?> composition) {
+    public void addComposition(ObjectRef<? extends ObjectId> composition) {
         this.compositions.add(composition);
     }
 
@@ -145,11 +150,11 @@ public class Ehr extends RMObject {
     }
 
     @Nullable
-    public List<ObjectRef<?>> getFolders() {
+    public List<ObjectRef<? extends ObjectId>> getFolders() {
         return folders;
     }
 
-    public void setFolders(@Nullable List<ObjectRef<?>> folders) {
+    public void setFolders(@Nullable List<ObjectRef<? extends ObjectId>> folders) {
         this.folders = folders;
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/Ehr.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/Ehr.java
@@ -9,11 +9,7 @@ import com.nedap.archie.rminfo.Invariant;
 import com.nedap.archie.rmutil.InvariantUtil;
 
 import javax.annotation.Nullable;
-import javax.xml.bind.annotation.XmlAccessType;
-import javax.xml.bind.annotation.XmlAccessorType;
-import javax.xml.bind.annotation.XmlElement;
-import javax.xml.bind.annotation.XmlRootElement;
-import javax.xml.bind.annotation.XmlType;
+import javax.xml.bind.annotation.*;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Objects;
@@ -23,7 +19,7 @@ import java.util.Objects;
  * in some kind of persistence framework. But it can be useful for transferring information
  * Created by pieter.bos on 08/07/16.
  */
-@XmlRootElement(name = "EHR")
+@XmlRootElement(name="EHR")
 @XmlType(name = "EHR", propOrder = {
         "systemId",
         "ehrId",
@@ -38,15 +34,15 @@ import java.util.Objects;
 @XmlAccessorType(XmlAccessType.FIELD)
 public class Ehr extends RMObject {
 
-    @XmlElement(name = "system_id")
+    @XmlElement(name="system_id")
     private HierObjectId systemId;
-    @XmlElement(name = "ehr_id")
+    @XmlElement(name="ehr_id")
     private HierObjectId ehrId;
 
     private List<ObjectRef<? extends ObjectId>> contributions = new ArrayList<>();
-    @XmlElement(name = "ehr_status")
+    @XmlElement(name="ehr_status")
     private ObjectRef<?> ehrStatus;
-    @XmlElement(name = "ehr_access")
+    @XmlElement(name="ehr_access")
     private ObjectRef<?> ehrAccess;
     @Nullable
     private List<ObjectRef<? extends ObjectId>> compositions = new ArrayList<>();
@@ -57,7 +53,7 @@ public class Ehr extends RMObject {
     @Nullable
     private List<ObjectRef<? extends ObjectId>> folders = new ArrayList<>();
 
-    @XmlElement(name = "time_created")
+    @XmlElement(name="time_created")
     private DvDateTime timeCreated;
 
     public Ehr() {

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/VersionedComposition.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/VersionedComposition.java
@@ -4,8 +4,8 @@ import com.nedap.archie.rm.changecontrol.VersionedObject;
 import com.nedap.archie.rm.composition.Composition;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
-import com.nedap.archie.rminfo.Invariant;
 
 /**
  * Created by pieter.bos on 08/07/16.
@@ -15,7 +15,7 @@ public class VersionedComposition extends VersionedObject<Composition> {
     public VersionedComposition() {
     }
 
-    public VersionedComposition(HierObjectId uid, ObjectRef<?> ownerId, DvDateTime timeCreated) {
+    public VersionedComposition(HierObjectId uid, ObjectRef<? extends ObjectId> ownerId, DvDateTime timeCreated) {
         super(uid, ownerId, timeCreated);
     }
 

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/VersionedEhrAccess.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/VersionedEhrAccess.java
@@ -3,6 +3,7 @@ package com.nedap.archie.rm.ehr;
 import com.nedap.archie.rm.changecontrol.VersionedObject;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 
 import javax.xml.bind.annotation.XmlAccessType;
@@ -19,7 +20,7 @@ public class VersionedEhrAccess extends VersionedObject<EhrAccess> {
     public VersionedEhrAccess() {
     }
 
-    public VersionedEhrAccess(HierObjectId uid, ObjectRef<?> ownerId, DvDateTime timeCreated) {
+    public VersionedEhrAccess(HierObjectId uid, ObjectRef<? extends ObjectId> ownerId, DvDateTime timeCreated) {
         super(uid, ownerId, timeCreated);
     }
 }

--- a/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/VersionedEhrStatus.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rm/ehr/VersionedEhrStatus.java
@@ -3,6 +3,7 @@ package com.nedap.archie.rm.ehr;
 import com.nedap.archie.rm.changecontrol.VersionedObject;
 import com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime;
 import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 
 import javax.xml.bind.annotation.XmlAccessType;
@@ -19,7 +20,7 @@ public class VersionedEhrStatus extends VersionedObject<EhrStatus> {
     public VersionedEhrStatus() {
     }
 
-    public VersionedEhrStatus(HierObjectId uid, ObjectRef<?> ownerId, DvDateTime timeCreated) {
+    public VersionedEhrStatus(HierObjectId uid, ObjectRef<? extends ObjectId> ownerId, DvDateTime timeCreated) {
         super(uid, ownerId, timeCreated);
     }
 }

--- a/openehr-rm/src/main/java/com/nedap/archie/rmutil/InvariantUtil.java
+++ b/openehr-rm/src/main/java/com/nedap/archie/rmutil/InvariantUtil.java
@@ -3,6 +3,7 @@ package com.nedap.archie.rmutil;
 import com.nedap.archie.rm.datatypes.CodePhrase;
 import com.nedap.archie.rm.datavalues.DvCodedText;
 import com.nedap.archie.rm.datavalues.DvText;
+import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import com.nedap.archie.terminology.OpenEHRTerminologyAccess;
 import com.nedap.archie.terminology.TermCode;
@@ -61,7 +62,7 @@ public class InvariantUtil {
         return true;
     }
 
-    public static boolean objectRefTypeEquals(List<ObjectRef<?>> refs, String type) {
+    public static boolean objectRefTypeEquals(List<ObjectRef<? extends ObjectId>> refs, String type) {
         if(refs == null) {
             return true;
         }

--- a/openehr-rm/src/test/java/com/nedap/archie/rm/changecontrol/ContributionTest.java
+++ b/openehr-rm/src/test/java/com/nedap/archie/rm/changecontrol/ContributionTest.java
@@ -1,0 +1,34 @@
+package com.nedap.archie.rm.changecontrol;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.nedap.archie.json.JacksonUtil;
+import com.nedap.archie.json.RMJacksonConfiguration;
+import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectRef;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class ContributionTest {
+
+    @Test
+    public void testJsonSerialization() throws IOException {
+        Contribution expected = new Contribution();
+
+        List<ObjectRef<?>> versions = new ArrayList<>();
+        versions.add(new ObjectRef<>(new HierObjectId("value"), "namespace", "type"));
+        expected.setVersions(versions);
+
+        StringWriter sw = new StringWriter();
+        ObjectMapper objectMapper = JacksonUtil.getObjectMapper(RMJacksonConfiguration.createStandardsCompliant());
+        objectMapper.writeValue(sw, expected);
+
+        Contribution actual = objectMapper.readValue(sw.toString(), Contribution.class);
+        assertEquals(expected, actual);
+    }
+}

--- a/openehr-rm/src/test/java/com/nedap/archie/rm/composition/ObservationTest.java
+++ b/openehr-rm/src/test/java/com/nedap/archie/rm/composition/ObservationTest.java
@@ -1,35 +1,29 @@
-package com.nedap.archie.rm.changecontrol;
+package com.nedap.archie.rm.composition;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.nedap.archie.json.JacksonUtil;
 import com.nedap.archie.json.RMJacksonConfiguration;
 import com.nedap.archie.rm.support.identification.HierObjectId;
-import com.nedap.archie.rm.support.identification.ObjectId;
 import com.nedap.archie.rm.support.identification.ObjectRef;
 import org.junit.Test;
 
 import java.io.IOException;
 import java.io.StringWriter;
-import java.util.ArrayList;
-import java.util.List;
 
 import static org.junit.Assert.assertEquals;
 
-public class ContributionTest {
+public class ObservationTest {
 
     @Test
     public void testJsonSerialization() throws IOException {
-        Contribution expected = new Contribution();
+        Observation expected = new Observation();
 
-        List<ObjectRef<? extends ObjectId>> versions = new ArrayList<>();
-        versions.add(new ObjectRef<>(new HierObjectId("value"), "namespace", "type"));
-        expected.setVersions(versions);
-
+        expected.setGuidelineId(new ObjectRef<>(new HierObjectId("value"), "namespace", "type"));
         StringWriter sw = new StringWriter();
         ObjectMapper objectMapper = JacksonUtil.getObjectMapper(RMJacksonConfiguration.createStandardsCompliant());
         objectMapper.writeValue(sw, expected);
 
-        Contribution actual = objectMapper.readValue(sw.toString(), Contribution.class);
+        Observation actual = objectMapper.readValue(sw.toString(), Observation.class);
         assertEquals(expected, actual);
     }
 }

--- a/openehr-rm/src/test/java/com/nedap/archie/rm/ehr/EhrTest.java
+++ b/openehr-rm/src/test/java/com/nedap/archie/rm/ehr/EhrTest.java
@@ -21,8 +21,16 @@ public class EhrTest {
         Ehr expected = new Ehr();
 
         List<ObjectRef<?>> contributions = new ArrayList<>();
-        contributions.add(new ObjectRef<>(new HierObjectId("value"), "namespace", "type"));
+        contributions.add(new ObjectRef<>(new HierObjectId("contribution_value"), "contribution_namespace", "contribution_type"));
         expected.setContributions(contributions);
+
+        List<ObjectRef<?>> compositions = new ArrayList<>();
+        compositions.add(new ObjectRef<>(new HierObjectId("composition_value"), "composition_namespace", "composition_type"));
+        expected.setCompositions(compositions);
+
+        List<ObjectRef<?>> folders = new ArrayList<>();
+        folders.add(new ObjectRef<>(new HierObjectId("folder_value"), "folder_namespace", "folder_type"));
+        expected.setFolders(folders);
 
         StringWriter sw = new StringWriter();
         ObjectMapper objectMapper = JacksonUtil.getObjectMapper(RMJacksonConfiguration.createStandardsCompliant());

--- a/openehr-rm/src/test/java/com/nedap/archie/rm/ehr/EhrTest.java
+++ b/openehr-rm/src/test/java/com/nedap/archie/rm/ehr/EhrTest.java
@@ -1,0 +1,34 @@
+package com.nedap.archie.rm.ehr;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.nedap.archie.json.JacksonUtil;
+import com.nedap.archie.json.RMJacksonConfiguration;
+import com.nedap.archie.rm.support.identification.HierObjectId;
+import com.nedap.archie.rm.support.identification.ObjectRef;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class EhrTest {
+
+    @Test
+    public void testJsonSerialization() throws IOException {
+        Ehr expected = new Ehr();
+
+        List<ObjectRef<?>> contributions = new ArrayList<>();
+        contributions.add(new ObjectRef<>(new HierObjectId("value"), "namespace", "type"));
+        expected.setContributions(contributions);
+
+        StringWriter sw = new StringWriter();
+        ObjectMapper objectMapper = JacksonUtil.getObjectMapper(RMJacksonConfiguration.createStandardsCompliant());
+        objectMapper.writeValue(sw, expected);
+
+        Ehr actual = objectMapper.readValue(sw.toString(), Ehr.class);
+        assertEquals(expected, actual);
+    }
+}

--- a/openehr-rm/src/test/java/com/nedap/archie/rm/ehr/EhrTest.java
+++ b/openehr-rm/src/test/java/com/nedap/archie/rm/ehr/EhrTest.java
@@ -19,6 +19,9 @@ public class EhrTest {
     @Test
     public void testJsonSerialization() throws IOException {
         Ehr expected = new Ehr();
+        expected.setEhrStatus(new ObjectRef<>(new HierObjectId("ehr_status_value"), "ehr_status_namespace", "ehr_status_type"));
+        expected.setEhrAccess(new ObjectRef<>(new HierObjectId("ehr_access_value"), "ehr_access_namespace", "ehr_access_type"));
+        expected.setDirectory(new ObjectRef<>(new HierObjectId("directory_value"), "directory_namespace", "directory_type"));
 
         List<ObjectRef<?>> contributions = new ArrayList<>();
         contributions.add(new ObjectRef<>(new HierObjectId("contribution_value"), "contribution_namespace", "contribution_type"));

